Add support for preprocessor directives to fparser2

reuterbal opened the pull request:
This adds support for preprocessor directives to fparser2 as discussed in #215 

It operates under the assumption that the code reduces to valid Fortran when preprocessor directives are left out. The implementation follows closely the description in the [C99 draft](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf#page=157), omitting only the `#pragma` statement.

Please mention any changes that you would like to be made.

reuterbal commented:
OK, apparently the section symbol used in the draft isn't ASCII... removed it from comments and docstring

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+    @staticmethod
+    def match(string):
+        if not string:
```
Again, docstring describing the method and a description of the arguments please. I'll not add this comment to any subsequent functions/classes with missing docstrings but please add them there too.

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+    def match(string):
+        '''Implements the matching for an include statement.
+
+        :param str string: the string to match with as an include statement.
```
Ah yes, docstrings like this :-)

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+        if not string:
+            return None
+
```
Previous matches do not have a blank line between the if test and line strip.

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+    def tostr(self):
+        '''
+        :return: this include_stmt as a string
```
full stop at the end of the description please.

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+            definition = definition[found.end():].strip()
+            return (name, parameter_list, definition)
+        definition = definition.strip()
+        # now that we now it doesn't have a parameter list, we can strip
```
now -> know

rupertford replied:
Should this comment be above the previous line  to make sense.

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+    def tostr(self):
+        if len(self.items) > 2:
+            return ('#define {}{} {}'.format(self.items[0], self.items[1],
```
Again `{0}` instead of `{}` if you don't mind, or can't convince me.

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+                       | (...)
+    '''
+
+    _regex = re.compile(r'\(\s*[A-Za-z_]\w*(\s*,\s*[A-Za-z_])*\s*\)')
```
Delete this one as the value of `_regex` is recomputed on the next line?

rupertford commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+    directive keyword (undef instead of define) we treat it separately.
+    '''
+
+    # There are no other classes. This is a simple string match.
```
This comment is not true?

rupertford commented on src/fparser/two/Fortran2003.py:
```diff
@@ -168,26 +168,35 @@ def restore_reader(self, reader):
-    list. Comment and/or include objects are added until a line that
-    is not a comment or include is found.
+def match_comment_or_include(reader):
+    '''Create a comment or include object from the current line.'''
```
document arguments please.

rupertford commented on src/fparser/two/Fortran2003.py:
```diff
@@ -341,7 +347,7 @@ def match(string):
         if not name:
             raise InternalError(
-                "Fotran2003.py:Include_Stmt:match Include_Filename should "
+                "Fortran2003.py:Include_Stmt:match Include_Filename should "
```
Thanks!

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+#!/usr/bin/env python
```
Again, please remove this and add copyright notice.

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+# First version created: Jan 2020
+
+import pytest
+from fparser.two.C99Preprocessor import (Cpp_If_Stmt, Cpp_Elif_Stmt,
```
I think I'd have a newline before Cpp_If_Stmt to allow more code per line.

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+from fparser.api import get_reader
+
+
+def test_if_stmt(f2003_create):
```
Use of f2003_create causes pylint to complain so we now suggest using the following instead:
`@pytest.mark.usefixtures("f2003_create")`

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+    '''Test that incorrectly formed #if statements raise exception'''
+    for line in [None, '', ' ', '#ifdfe', '#if', '#ifdef', '#ifdef two macros']:
+        with pytest.raises(NoMatchError) as excinfo:
+            print(line)
```
please remove print

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+
+def test_incorrect_if_stmt(f2003_create):
+    '''Test that incorrectly formed #if statements raise exception'''
+    for line in [None, '', ' ', '#ifdfe', '#if', '#ifdef', '#ifdef two macros']:
```
I suggest changing this to use `pytest.mark.parametrize(...)`

rupertford replied:
And for subsequent code wit the same structure.

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,524 @@
+
+
+def test_parse_define_outside_subroutine(f2003_create):
+    f2003_parser = ParserFactory().create(std='f2003')
```
No need for `f2003_create` if you create the parser here.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -564,7 +570,13 @@ def match(startcls, subclasses, endcls, reader,
         # Comments and Include statements are always valid sub-classes
-        classes = subclasses + [Comment, Include_Stmt]
+        # Preprocessor directives are always valid sub-classes
+        classes = subclasses + [Comment, Include_Stmt, Cpp_If_Stmt,
```
Could the class list from C99*.py be imported to provide the list of cpp classes?

reuterbal replied:
I was in fact using `get_module_classes` from parser.py for this in an earlier version and was under the impression that this slowed down the parser significantly (as the list is re-built quite often). I'll have a look again.

reuterbal replied:
I have now put a list with the relevant class names in C99Preprocessor.py. This list then used here as well as in `match_cpp_directive`. This way, the overhead seems to be non measurable.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -669,6 +681,21 @@ def match(startcls, subclasses, endcls, reader,
                     i = 1
                 if isinstance(obj, End_Select_Stmt):
                     enable_case_construct_hook = False
+            #
```
remove the commented-out code please.

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -508,6 +508,7 @@ def match(startcls, subclasses, endcls, reader,
               enable_where_construct_hook=False,
               enable_select_type_construct_hook=False,
               enable_case_construct_hook=False,
+              # enable_cpp_if_construct_hook=False,
```
remove please

rupertford commented on src/fparser/two/utils.py:
```diff
@@ -527,20 +528,25 @@ def match(startcls, subclasses, endcls, reader,
         :param bool enable_where_construct_hook: TBD
         :param bool enable_select_type_construct_hook: TBD
         :param bool enable_case_construct_hook: TBD
+        :param bool enable_cpp_if_construct_hook: TBD
```
remove please

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+#!/usr/bin/env python
+
+# All rights reserved.
```
Sorry, missed that. I copied the header from Fortran2003.py and removed this to ask someone what I am supposed to use - and of course forgot about it.

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+import re
+
+from fparser.two import pattern_tools as pattern
+from fparser.two.utils import (Base, StringBase, InternalError)
```
It is actually superfluous in this case - I guess I had more imports in that statement in an earlier stage, which required a line break, and never removed the parenthesis when no longer needed.

It is accepted standard (introduced in [PEP328](https://www.python.org/dev/peps/pep-0328/)) and was partially born out of PEP8's [preference of parentheses instead of backslash](https://www.python.org/dev/peps/pep-0008/#maximum-line-length) for line continuation:
> The preferred way of wrapping long lines is by using Python's implied line continuation inside parentheses, brackets and braces. Long lines can be broken over multiple lines by wrapping expressions in parentheses. These should be used in preference to using a backslash for line continuation.

Personally, I don't always see it as advantageous as it mandates indentation on following lines to align with the opening parenthesis, which limits the useful line length significantly when the import path is long.

reuterbal commented:
Thanks a lot for your review and the suggested changes!
I'll work through as much as I can now and let you know once I think it's worth having another look

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+        return (kind, rhs)
+
+    def tostr(self):
+        return ('#{} {}'.format(self.items[0], self.items[1]))
```
My only motivation is lazyness: saves a few characters and avoids the hassle of renumbering if something is to be added inbetween (of course not to be expected in this case). For the latter one could then also write `'{0} {2} {1}'` but I find that much more confusing than not putting any position numbers in the first place.

reuterbal replied:
(Not meant to be an attempt to convince you: I'll happily add the position numbers)

rupertford replied:
Thanks.

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+
+#
+# Below construct for if-elif-else-endif blocks of preprocessor macros works
```
Agreed.

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+    subclass_names = []
+    use_names = ['Cpp_Macro_Identifier']
+
+    _regex = re.compile(r"#\s*(ifdef|ifndef|if)\b")  # 'if' last, order matters
```
This comment was actually a leftover from @martin-schlipf's work. Order would matter if word ending (via `\b`) wasn't explicitly required after the group of alternatives, since then `if` would also match `ifdef`. As this is not the case, I removed the comment.

reuterbal commented on src/fparser/two/C99Preprocessor.py:
```diff
@@ -0,0 +1,561 @@
+
+def match_cpp_directive(reader):
+    '''Create single-line C99 preprocessor directive object from the current
+    line. Omit if-construct and its members that are dealt with separately.'''
```
Done for every routine that I could find ;-)

reuterbal commented:
@rupertford I tried to address all of your requested changes. 

I have marked most of your comments as resolved after having worked on that bit simply to keep track of what remains to be done (Not sure if this is just a personal setting or if they appear like that for you as well). It may not necessarily mean that the raised point is resolved to your satisfaction ;-)

When you find time I'd be glad if you could have another look and come back with some further remarks.

rupertford commented:
Thanks @reuterbal. I've changed the status to ready-for-review and will review when I next get the chance. For the record, our policy is that the reviewer marks things resolved :-). We would be happy to give you access to the repo if you would prefer to work here rather than on a fork.

reuterbal commented:
Sorry, I un-resolved them again...

I would not mind repository access but it's also not a problem to work on a fork. It doesn't really change much for me. Any changes that I have to make I will send your way as a PR anyways and there will be an issue about that beforehand (unless it's something very tiny like #241).
I have an additional branch `loki_integration` on my fork which I only use to merge branches that are not yet in fparser's master but are necessary for my downstream work. This makes it easier for me in the CI testing.

reuterbal commented:
Thanks for looking through it again. Code duplication is a fair point that I did not think about since most of the matching was already implemented by @martin-schlipf and I changed only a few bits. 
One concern would probably be the fact that preprocessor directives are case-sensitive (I think) and everything in utils.py is not. The question is, whether to make case sensitive matching an optional feature in WORDClsBase etc. or a very similar base class should be added?

Looking through the relevant matching routines and leaving case sensitivity aside, I think the following could be done:
- Match the `#` explicitly and use one of the classes from utils.py to match the remainder of the line
- `Cpp_If_Stmt`, `Cpp_Elif_Stmt`: Introduce a new object for the right hand side (if it's not an if(n)def) instead of keeping it as a string and use `WORDClsBase` to do the matching (similar to, e.g., `Intrinsic_Type_Spec`).
- `Cpp_Else_Stmt`, `Cpp_Endif_Stmt`: Use `WORDClsBase` without cls
- `Cpp_Undef_Stmt`: Use `WORDClsBase` with 
- `Cpp_Include_Stmt`: Keep as is (similar to `Include_Stmt`)
- `Cpp_Line_Stmt`: Use `WORDClsBase` with a new right hand side class (either a new one matching closely the standard or a rather loose matching using the one for if/elif)
- `Cpp_Macro_Stmt`: This one could prove a bit more difficult to get the handling of white spaces and the different meanings of those right. Could maybe work with a clever regex and `STRINGBase`. Otherwise I would just keep it as is.
- `Cpp_Macro_Identifier_List`: Put the regex into pattern_tools.py and use `StringBase`. I also noticed that the regex is not yet fully correct as it allows for the variadic argument dots in any position (not just the last).
- `Cpp_Error_Stmt`, `Cpp_Warning_Stmt`: Use `WORDClsBase` with a new cls.
- `Cpp_Null_Stmt`: Keep as is.

Thus, in principal it should be possible but the question is how to take care of case sensitivity.

rupertford commented:
Yes, good point about case insensitive matching, I forgot about that.

The simplest thing would probably be to add an an `ignore_case=True` to the `match` method in `WORDClsBase()` and then set that to `False` in the new CPP matches. I think the only code change required would be to `if line[:len(keyword)].upper() != keyword.upper():`. I would accept a new issue for adding tests to WORDClsBase rather than making you add new tests for it (unless you want to).

A nicer way would be to do a case sensitive match and then subclass this for a case insensitive match and use `WordClsBase` for case sensitive and `WORDClsBase` for case insensitive, following the `StringBase` and `STRINGBase` naming convention. However that would probably be a separate PR.

reuterbal commented:
If it's OK I would go with the `ignore_case=True` option to `WORDClsBase` for now and opening an issue to remove this and implement the suggested hierarchy instead -- or at least adding the relevant tests. I might find time inbetween to do that but I would like to get on with the other issues first.

rupertford commented:
Fine by me.

reuterbal commented:
Turns out I can even make it without changes to `WORDClsBase` because with `Pattern` matching the case sensitivity is defined in the pattern. However, adding a `WordClsBase` and using that as base class is probably still desirable for clarity. 
I have now rewritten almost all of the matching routines using either WORDClsBase or StringBase as base class. Cpp_Macro_Stmt is kept as is because the matching needs to properly take care of the different behaviour with or without whitespace behind the identifier.
For now I have defined the patterns as class properties instead of adding them to `pattern_tools.py` to avoid cluttering that file with too many of those. But I could also put them there with a `cpp_` prefix if you prefer?
I'll add some documentation tomorrow - sorry, I only now realized that you didn't refer to the docstrings by that.

reuterbal commented:
Thanks for merging the other PR. I have updated this branch to current master and added a check that short circuits the match routines for CPP directives if `reader` has not returned a `CppDirective` object.
Documentation was added in the fparser2 section as well as the developer guide. I hope that's what you had in mind. I also tried to update the docstrings to match the style discussed in the other PR.
I think this is now ready for another look.

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,569 @@
+
+
+@pytest.mark.usefixtures("f2003_create")
+@pytest.mark.parametrize('line, ref', [
```
Almost all of the inputs and outputs are the same with parametrize. You could either split the tests into identical ones and ones with extra spaces and then for the identical just pass in the reference, or you could just pass in the original for all of them and then check with e.g. `' '.join(line.split())`. Either way there is no need for both `line` and `ref` for most (or all) of the tests.

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,569 @@
+def test_parse_define_outside_subroutine():
+    f2003_parser = ParserFactory().create(std='f2003')
+
+    code = '#define MACRO\nSUBROUTINE FOO\ncall sub()\nEND SUBROUTINE FOO\n'
```
You could just make the input code the same as ref then only have one of them?

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,569 @@
+def test_parse_empty_ifdef():
+    f2003_parser = ParserFactory().create(std='f2003')
+
+    code = '''
```
For this test and subsequent tests could you reduce the ammount of code by making the input the same as the output, or the input a stripped version of the output if you prefer, rather than replicating the code?

rupertford commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,569 @@
+
+
+@pytest.mark.usefixtures("f2003_create")
+@pytest.mark.parametrize('line, ref', [
```
Again, most of these `line` and `ref` values are the same. Could you simplify?

reuterbal commented on doc/developers_guide.rst:
```diff
@@ -726,6 +726,73 @@ any 'program units'. This completes all valid locations for include
+Preprocessing Directives
+++++++++++++++++++++++++
+
+fparser2 retaines preprocessing directives as nodes in the parse tree
```
Thanks

reuterbal commented on doc/developers_guide.rst:
```diff
@@ -726,6 +726,73 @@ any 'program units'. This completes all valid locations for include
+`C99Preprocessor.py` as a number of classes that have names with the
+prefix `Cpp_`. This allows fparser2 to parse code successfully that
+contains preprocessing directives but reduces to valid Fortran if the
+directives are omitted.preprocessing directives are represented by a
```
Uuh, no idea where that came from.

reuterbal commented on src/fparser/two/Fortran2003.py:
```diff
@@ -167,26 +167,46 @@ def restore_reader(self, reader):
                   or \
                   :py:class:`fparser.common.readfortran.FortranStringReader`
+    :returns: A comment or include object if found, otherwise `None`.
+    :rtype: :py:class:`fparser.two.Fortran2003.Comment` or \
```
No, those are in fact `Comment` or `Include_Stmt`

rupertford replied:
Oh yes, oops.

reuterbal commented:
No need to apologize, really. I appreciate the scrutiny.
This should address all of the remarks - sorry from my side for repeating mistakes that were mentioned before.
Code duplication in the tests should be significantly reduced and I hope this is now on an acceptable level.
Codecov reporting is somewhat off for me at the moment. It gave <100% for C99Preprocessor after I added the check for `CppDirective`, I fixed this by combining `return None` cases and now the reporting is screwed up. However, looking at the details on codecov.io doesn't give any red line. Maybe it'll repair the message itself later?

rupertford commented on doc/developers_guide.rst:
```diff
@@ -726,6 +726,73 @@ any 'program units'. This completes all valid locations for include
+by a dedicated class `CppDirective`, which is a subclass of `Line`.
+This allows to detect directives early on and attempt to match their
+classes only for source lines of that type. Matching of directives
+is done in the same place as include statements to make sure that they
```
performed not preformed of course :-)

reuterbal commented on src/fparser/two/tests/test_c99preprocessor.py:
```diff
@@ -0,0 +1,426 @@
+from fparser.api import get_reader
+
+
+@pytest.fixture(scope='module', name='f2003_parser')
```
The fixture in conftest.py creates the parser but does not return it. However, in a few tests I'm parsing longer pieces of code to check that directives can appear everywhere in a source file, i.e., I do need that parser, which is why I added that separate fixture here. I could also change the fixture in conftest.py such that it does return the parser, if you prefer?

rupertford replied:
Ah yes, good point, I missed that. If you could move the fixture you have created into conftest.py so it is available to other tests in the future that would be great. We can decide whether to merge them into one fixture at a later date (I would be in favour but let's minimise the changes you have to make in this PR).

reuterbal commented:
Many thanks, happy to hear that :)
Just for clarity: I should add my name in the relevant places in `doc/conf.py`and `doc/license.rst`? What ruleset do you follow regarding order of names? (Out of instinct I would have gone for 3rd place?)

rupertford commented:
> Many thanks, happy to hear that :)
> Just for clarity: I should add my name in the relevant places in `doc/conf.py`and `doc/license.rst`? What ruleset do you follow regarding order of names? (Out of instinct I would have gone for 3rd place?)

I've just talked to @arporter about this. We agree that you should go 3rd if you are OK with that. We don't currently have an official policy, it was just the order in which people contributed to the documentation (in time) + Pearu, so it is at least consistent with that informal policy. We agree there should be an official policy but suggest we park that decision for the time being.

And, yes, the files you mention are the correct ones. If you wouldn't mind updating the copyright date to 2020 as well that would be great (but the PR will still be approved if you don't :-)).

reuterbal commented:
I have moved the fixture and updated the author list (changed copyright dates wherever I've seen some). I don't care much about the order but wanted to avoid tipping on any toes.

Many thanks again for taking the time to work through this big change. This is a very crucial feature for our downstream tools and I'm very happy that it makes its way into master.

p-vitt commented:
Just for clarification: Is it correct, that only C preprocessor directives are supported? I'm currently stumbling over [CoCo](http://www.daniellnagle.com/index.html) directives starting with `?? macroname...`, but it seems that those can't be handled with this PR out of the box, right?

rupertford commented:
> Just for clarification: Is it correct, that only C preprocessor directives are supported? I'm currently stumbling over [CoCo](http://www.daniellnagle.com/index.html) directives starting with `?? macroname...`, but it seems that those can't be handled with this PR out of the box, right?

Yes that's right . This PR added support for cpp directives. Do you need to reason about the CoCo directives or would it be OK if they were removed?

rupertford commented:
> No, for my current task it would be sufficient to ignore them. But I fear that this won't be as straight forward as there are lines starting with ?? but also macros within code lines like:
> 
> ```
> value = ?MACRONAME?(here, come, some, arguments)
> ```

Ah. Yes, that is a problem. Presumably you are not wanting/able to pre-process the code first. I don't know much about CoCo but pre-processor syntax and Fortran syntax are not typically designed to be mixed. We could follow the approach taken by the cpp support but it would have the same restriction that the code without the directives was valid Fortran (which would include, not allowing inline macro code unless the macros always aligned with Fortran tokens - then we could potentially extend the associated rule).

p-vitt commented:
Well, I want to avoid operating on intermediate files. But I also want to create a working script, thus I'll go for the preprocessed files. Trying to get fparser to be able to cope with CoCo directives would take considerable effort, I presume. Nevertheless, thanks for your fast response :)

rupertford commented:
No problem. Getting fparser to cope with full line CoCo directives should be relatively simple, so if that would be useful we could look at adding that. Adding understanding about the Coco syntax of full line directives would be feasible, but a reasonable amount of effort, but it sounds like you don't need this. Adding support for inline macros would be considerable effort I suspect and might only work in certain circumstances.

p-vitt commented:
Exactly, for my current problem I don't need to understand CoCo syntax. Additionally, we are making plans to switch to Fypp (in the not so near future), which would introduce it's own specialties. Thanks for clarifications and support. :)
